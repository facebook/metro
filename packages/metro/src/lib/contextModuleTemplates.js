/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import * as path from 'path';
import type {ContextMode} from '../ModuleGraph/worker/collectDependencies';

function createFileMap(
  modulePath: string,
  files: string[],
  processModule: (moduleId: string) => string,
): string {
  let mapString = '\n';

  files
    .slice()
    // Sort for deterministic output
    .sort()
    .forEach(file => {
      let filePath = path.relative(modulePath, file);

      // NOTE(EvanBacon): I'd prefer we prevent the ability for a module to require itself (`require.context('./')`)
      // but Webpack allows this, keeping it here provides better parity between bundlers.

      // Ensure relative file paths start with `./` so they match the
      // patterns (filters) used to include them.
      if (!filePath.startsWith('.')) {
        filePath = `.${path.sep}` + filePath;
      }
      const key = JSON.stringify(filePath);
      // NOTE(EvanBacon): Webpack uses `require.resolve` in order to load modules on demand,
      // Metro doesn't have this functionality so it will use getters instead. Modules need to
      // be loaded on demand because if we imported directly then users would get errors from importing
      // a file without exports as soon as they create a new file and the context module is updated.

      // NOTE: The values are set to `enumerable` so the `context.keys()` method works as expected.
      mapString += `  ${key}: { enumerable: true, get() { return ${processModule(
        file,
      )}; } },\n`;
    });
  return `Object.defineProperties({}, {${mapString}})`;
}

function getEmptyContextModuleTemplate(modulePath: string): string {
  return `
function metroEmptyContext(request) {
  let e = new Error('No modules in context');
  e.code = 'MODULE_NOT_FOUND';
  throw e;
}

// Return the keys that can be resolved.
metroEmptyContext.keys = () => ([]);

// Return the module identifier for a user request.
metroEmptyContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroEmptyContext;`;
}

function getLoadableContextModuleTemplate(
  modulePath: string,
  files: string[],
  importSyntax: string,
  getContextTemplate: string,
): string {
  return `// All of the requested modules are loaded behind enumerable getters.
const map = ${createFileMap(
    modulePath,
    files,
    moduleId => `${importSyntax}(${JSON.stringify(moduleId)})`,
  )};
  
function metroContext(request) {
  ${getContextTemplate}
}
  
// Return the keys that can be resolved.
metroContext.keys = function metroContextKeys() {
  return Object.keys(map);
};

// Return the module identifier for a user request.
metroContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroContext;`;
}

/**
 * Generate a context module as a virtual file string.
 *
 * @prop {ContextMode} mode indicates how the modules should be loaded.
 * @prop {string} modulePath virtual file path for the virtual module. Example: `require.context('./src')` -> `'/path/to/project/src'`.
 * @prop {string[]} files list of absolute file paths that must be exported from the context module. Example: `['/path/to/project/src/index.js']`.
 *
 * @returns a string representing a context module (virtual file contents).
 */
export function getContextModuleTemplate(
  mode: ContextMode,
  modulePath: string,
  files: string[],
): string {
  if (!files.length) {
    return getEmptyContextModuleTemplate(modulePath);
  }
  switch (mode) {
    case 'eager':
      return getLoadableContextModuleTemplate(
        modulePath,
        files,
        // NOTE(EvanBacon): It's unclear if we should use `import` or `require` here so sticking
        // with the more stable option (`require`) for now.
        'require',
        [
          '  // Here Promise.resolve().then() is used instead of new Promise() to prevent',
          '  // uncaught exception popping up in devtools',
          '  return Promise.resolve().then(() => map[request]);',
        ].join('\n'),
      );
    case 'sync':
      return getLoadableContextModuleTemplate(
        modulePath,
        files,
        'require',
        '  return map[request];',
      );
    case 'lazy':
    case 'lazy-once':
      return getLoadableContextModuleTemplate(
        modulePath,
        files,
        'import',
        '  return map[request];',
      );
    default:
      throw new Error(`Metro context mode "${mode}" is unimplemented`);
  }
}
