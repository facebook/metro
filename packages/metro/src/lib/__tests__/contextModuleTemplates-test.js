/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall react_native
 */

import {getContextModuleTemplate} from '../contextModuleTemplates';

describe('getContextModuleTemplate', () => {
  test('creates a sync template', () => {
    const template = getContextModuleTemplate('sync', '/path/to/project/src', [
      '/path/to/project/src/foo.js',
    ]);
    expect(template).toMatch(/foo\.js/);
    expect(template).toMatchSnapshot();
  });
  test('creates an empty template', () => {
    const template = getContextModuleTemplate(
      'sync',
      '/path/to/project/src',
      [],
    );
    expect(template).toMatch(/MODULE_NOT_FOUND/);
    expect(template).toMatchSnapshot();
  });
  test('creates an eager template', () => {
    const template = getContextModuleTemplate('eager', '/path/to/project/src', [
      '/path/to/project/src/foo.js',
    ]);
    expect(template).toMatchSnapshot();
  });
  test('creates a lazy template', () => {
    const template = getContextModuleTemplate('lazy', '/path/to/project/src', [
      '/path/to/project/src/foo.js',
    ]);
    expect(template).toMatchSnapshot();
  });
  test('creates a lazy-once template', () => {
    const template = getContextModuleTemplate(
      'lazy-once',
      '/path/to/project/src',
      ['/path/to/project/src/foo.js', '/path/to/project/src/another/bar.js'],
    );

    expect(template).toMatchSnapshot();
  });

  test('creates posix paths on windows for sync template', () => {
    jest.resetModules();
    jest.mock('path', () => jest.requireActual('path').win32);
    const {
      getContextModuleTemplate: getWindowsTemplate,
    } = require('../contextModuleTemplates');
    const template = getWindowsTemplate('sync', 'c:/path/to/project/src', [
      'C:\\path\\to\\project\\src\\foo.js',
    ]);
    expect(template).toMatch(/foo\.js/);
    expect(template).toMatchSnapshot();
  });
});
