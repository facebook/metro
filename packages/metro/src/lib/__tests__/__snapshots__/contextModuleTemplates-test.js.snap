// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getContextModuleTemplate creates a lazy template 1`] = `
"// All of the requested modules are loaded behind enumerable getters.
const map = Object.defineProperties({}, {
  \\"./foo.js\\": { enumerable: true, get() { return import(\\"/path/to/project/src/foo.js\\"); } },
});

function metroContext(request) {
    return map[request];
}

// Return the keys that can be resolved.
metroContext.keys = function metroContextKeys() {
  return Object.keys(map);
};

// Return the module identifier for a user request.
metroContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroContext;"
`;

exports[`getContextModuleTemplate creates a lazy-once template 1`] = `
"// All of the requested modules are loaded behind enumerable getters.
const map = Object.defineProperties({}, {
  \\"./another/bar.js\\": { enumerable: true, get() { return import(\\"/path/to/project/src/another/bar.js\\"); } },
  \\"./foo.js\\": { enumerable: true, get() { return import(\\"/path/to/project/src/foo.js\\"); } },
});

function metroContext(request) {
    return map[request];
}

// Return the keys that can be resolved.
metroContext.keys = function metroContextKeys() {
  return Object.keys(map);
};

// Return the module identifier for a user request.
metroContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroContext;"
`;

exports[`getContextModuleTemplate creates a sync template 1`] = `
"// All of the requested modules are loaded behind enumerable getters.
const map = Object.defineProperties({}, {
  \\"./foo.js\\": { enumerable: true, get() { return require(\\"/path/to/project/src/foo.js\\"); } },
});

function metroContext(request) {
    return map[request];
}

// Return the keys that can be resolved.
metroContext.keys = function metroContextKeys() {
  return Object.keys(map);
};

// Return the module identifier for a user request.
metroContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroContext;"
`;

exports[`getContextModuleTemplate creates an eager template 1`] = `
"// All of the requested modules are loaded behind enumerable getters.
const map = Object.defineProperties({}, {
  \\"./foo.js\\": { enumerable: true, get() { return require(\\"/path/to/project/src/foo.js\\"); } },
});

function metroContext(request) {
    // Here Promise.resolve().then() is used instead of new Promise() to prevent
  // uncaught exception popping up in devtools
  return Promise.resolve().then(() => map[request]);
}

// Return the keys that can be resolved.
metroContext.keys = function metroContextKeys() {
  return Object.keys(map);
};

// Return the module identifier for a user request.
metroContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroContext;"
`;

exports[`getContextModuleTemplate creates an empty template 1`] = `
"
function metroEmptyContext(request) {
  let e = new Error('No modules in context');
  e.code = 'MODULE_NOT_FOUND';
  throw e;
}

// Return the keys that can be resolved.
metroEmptyContext.keys = () => ([]);

// Return the module identifier for a user request.
metroEmptyContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroEmptyContext;"
`;

exports[`getContextModuleTemplate creates posix paths on windows for sync template 1`] = `
"// All of the requested modules are loaded behind enumerable getters.
const map = Object.defineProperties({}, {
  \\"./foo.js\\": { enumerable: true, get() { return require(\\"C:\\\\\\\\path\\\\\\\\to\\\\\\\\project\\\\\\\\src\\\\\\\\foo.js\\"); } },
});

function metroContext(request) {
    return map[request];
}

// Return the keys that can be resolved.
metroContext.keys = function metroContextKeys() {
  return Object.keys(map);
};

// Return the module identifier for a user request.
metroContext.resolve = function metroContextResolve(request) {
  throw new Error('Unimplemented Metro module context functionality');
}

module.exports = metroContext;"
`;
